compilers/concrete-compiler/compiler/lib/Dialect/RT/Transforms/HoistAwaitFuturePass.cpp

```diff
@@ -0,0 +1,259 @@
+      auto &newOperations = newForallOp.getBody()->getOperations();
+      mlir::Block *oldBody = forallOp.getBody();
+
+      newOperations.splice(newOperations.begin(), oldBody->getOperations(),
+                           oldBody->begin(), std::prev(oldBody->end()));
+
+      // Wrap future in a tensor of one element, so that it can be
+      // stored in the new shared output tensor of futures using
+      // `tensor.parallel_insert_slice`
+      rewriter.setInsertionPointAfter(awaitFutureOp);
+      mlir::Value futureAsTensor =
+          rewriter.create<mlir::tensor::FromElementsOp>(
+              awaitFutureOp.getLoc(),
+              mlir::ValueRange{awaitFutureOp.getInput()});
+
+      // Move all operations from the old `scf.forall.in_parallel`
+      // terminator to the new one
+      mlir::scf::InParallelOp oldTerminator = forallOp.getTerminator();
+      mlir::scf::InParallelOp newTerminator = newForallOp.getTerminator();
+
+      mlir::Block::OpListType &oldTerminatorOps =
+          oldTerminator.getRegion().getBlocks().begin()->getOperations();
+      mlir::Block::OpListType &newTerminatorOps =
+          newTerminator.getRegion().getBlocks().begin()->getOperations();
+
+      newTerminatorOps.splice(newTerminatorOps.begin(), oldTerminatorOps,
+                              oldTerminatorOps.begin(), oldTerminatorOps.end());
+
+      // Remap IVs and out args
+      for (auto [oldIV, newIV] : llvm::zip(forallOp.getInductionVars(),
+                                           newForallOp.getInductionVars())) {
+        oldIV.replaceAllUsesWith(newIV);
+      }
+
+      {
+        size_t offs = 0;
+        for (auto it : llvm::enumerate(forallOp.getRegionOutArgs())) {
+          mlir::Value oldRegionOutArg = it.value();
+
+          if (oldRegionOutArg != dst) {
+            oldRegionOutArg.replaceAllUsesWith(
+                newForallOp.getRegionOutArgs()[it.index() - offs]);
+          } else {
+            offs++;
+          }
+        }
+      }
+
+      // Create new `tensor.parallel_inset_slice` operation inserting
+      // the future into the tensor of futures
+      llvm::SmallVector<mlir::OpFoldResult> ones(tripCounts.size(),
+                                                 rewriter.getI64IntegerAttr(1));
+
+      mlir::Value tensorOfFuturesRegionOutArg =
+          newForallOp.getRegionOutArgs().back();
+
+      mlir::ImplicitLocOpBuilder ilob(parallelInsertSliceOp.getLoc(), rewriter);
+
+      rewriter.setInsertionPointAfter(parallelInsertSliceOp);
+      rewriter.create<mlir::tensor::ParallelInsertSliceOp>(
+          parallelInsertSliceOp.getLoc(), futureAsTensor,
+          tensorOfFuturesRegionOutArg,
+          mlir::getAsOpFoldResult(mlir::concretelang::normalizeInductionVars(
+              ilob, newForallOp.getInductionVars(),
+              newForallOp.getMixedLowerBound(), newForallOp.getMixedStep())),
+          ones, ones);
+
+      // Create a new forall loop, that invoked `RT.await_future` on
```

Comment:
Definitely will want this to be a task in the future... Avoids a lot of synchronization.

Reply:
Agreed. Typo has been fixed.